processing/vector-indexes.mdx

```diff
@@ -0,0 +1,185 @@
+
+## Parameters
+
+| Parameter | Description | Valid for |
+| :--- | :--- | :--- |
+| `distance_type` | Distance metric to use: `l2`, `cosine`, `l1`, or `inner_product` | FLAT, HNSW |
+| `m` | Maximum number of connections per node | HNSW |
+| `ef_construction` | Size of dynamic candidate list during construction | HNSW |
+| `max_level` | Maximum level of the HNSW graph | HNSW |
+| `ef_search` | Controls the accuracy/speed trade-off for nearest neighbor searches. Can be set via the session variable `batch_hnsw_ef_search` (default 40). | HNSW |
+
+## Vector distance operators
+
+RisingWave provides specialized operators for calculating vector distances:
+
+| Operator | Function | Description |
+| :--- | :--- | :--- |
+| `<->` | `l2_distance()` | Euclidean (L2) distance |
+| `<=>` | `cosine_distance()` | Cosine distance |
+| `<+>` | `l1_distance()` | Manhattan (L1) distance |
+| `<#>` | Negative inner product | Negative inner product distance |
+
+## Vector similarity search
+
+Use vector distance operators with `ORDER BY` and `LIMIT` to perform similarity search:
+
+```sql
+-- Find the 5 most similar items using L2 distance
+SELECT * FROM items 
+ORDER BY embedding <-> '[3,1,2]' 
+LIMIT 5;
+
+-- Find similar items using cosine distance
+SELECT id, name FROM items 
+ORDER BY embedding <=> '[0.5, 0.3, 0.2]' 
+LIMIT 10;
+```
+
+## Vector data type, operators, functions
+
+RisingWave provides a `vector(n)` data type where `n` specifies the dimension of the vector.
+
+```sql
+-- Create a table with vector columns
+CREATE TABLE items (
+    id INT PRIMARY KEY,
+    name STRING,
+    embedding vector(3)
+) APPEND ONLY;
+```
+
+RisingWave also provides various operators and functions for working with vectors. See [Vector data type](/sql/data-types/vector) for details.
+
+## Vector indexes on function expressions
+
+You can create vector indexes on function expressions instead of raw columns. This allows you to avoid storing a separate vector column, saving storage and reducing maintenance costs.
+
+1. Create or alter the table to include the input column
+
+```sql
+ALTER TABLE items ADD COLUMN description STRING;
+```
+
+2. Define the user-defined function (UDF)
+
+```sql
+CREATE FUNCTION get_embedding(string) RETURNS VECTOR(128) LANGUAGE SQL AS $$
+SELECT openai_embedding('{"model": <EMBEDDING_MODEL_NAME>, "api_key": <API_KEY>}'::jsonb, $1)::vector(128);
+$$;
+```
+
+3. Create the vector index on the function expression
+
+```sql
+CREATE INDEX idx_embedding_func ON items
+USING FLAT (get_embedding(description))
+INCLUDE(description)
+WITH (distance_type = 'l2');
+```
+
+In this example, `get_embedding(description)` is used as the index expression.
+
+This approach avoids materializing a separate vector column in the table, which reduces storage costs and keeps the table schema simpler.
+
+## Examples
+
+### Basic vector similarity search
+
+```sql
+-- Create table with vector data
+CREATE TABLE products (
+    id INT PRIMARY KEY,
+    name STRING,
```

Comment:
Should include the `description` column in the table. The insert statements below should also be updated to include the extra `description` column.